mithril-aggregator/src/database/provider/signed_entity.rs

```diff
@@ -0,0 +1,593 @@
+use std::sync::Arc;
+
+use sqlite::{Connection, Value};
+
+use async_trait::async_trait;
+
+use mithril_common::{
+    entities::{SignedEntityType, Snapshot},
+    sqlite::{
+        EntityCursor, HydrationError, Projection, Provider, SourceAlias, SqLiteEntity,
+        WhereCondition,
+    },
+    store::adapter::{AdapterError, StoreAdapter},
+};
+
+use mithril_common::StdError;
+use tokio::sync::Mutex;
+
+/// SignedEntity record is the representation of a stored signed_entity.
+#[derive(Debug, PartialEq, Clone)]
+pub struct SignedEntityRecord {
+    /// Signed entity id.
+    signed_entity_id: String,
+
+    /// Signed entity type.
+    signed_entity_type: SignedEntityType,
+
+    /// Certificate id for this signed entity.
+    certificate_id: String,
+
+    /// Raw signed entity (in JSON format).
+    pub entity: String,
+
+    /// Date and time when the signed_entity was created
+    created_at: String,
+}
+
+impl From<Snapshot> for SignedEntityRecord {
+    fn from(other: Snapshot) -> Self {
+        let entity = serde_json::to_string(&other).unwrap();
+        SignedEntityRecord {
+            signed_entity_id: other.digest,
+            signed_entity_type: SignedEntityType::CardanoImmutableFilesFull(other.beacon),
+            certificate_id: other.certificate_hash,
+            entity,
+            created_at: other.created_at,
+        }
+    }
+}
+
+impl From<SignedEntityRecord> for Snapshot {
+    fn from(other: SignedEntityRecord) -> Snapshot {
+        serde_json::from_str(&other.entity).unwrap()
+    }
+}
+
+impl SqLiteEntity for SignedEntityRecord {
+    fn hydrate(row: sqlite::Row) -> Result<Self, HydrationError>
+    where
+        Self: Sized,
+    {
+        let signed_entity_id = row.get::<String, _>(0);
+        let signed_entity_type_id_int = row.get::<i64, _>(1);
+        let certificate_id = row.get::<String, _>(2);
+        let beacon_str = row.get::<String, _>(3);
+        let entity_str = row.get::<String, _>(4);
+        let created_at = row.get::<String, _>(5);
+
+        let signed_entity_record = Self {
+            signed_entity_id,
+            signed_entity_type: SignedEntityType::hydrate(
+                signed_entity_type_id_int.try_into().map_err(|e| {
+                    HydrationError::InvalidData(format!(
+                        "Could not cast i64 ({signed_entity_type_id_int}) to u64. Error: '{e}'"
+                    ))
+                })?,
+                &beacon_str,
+            )
+            .map_err(|e| {
```

Comment:
it already returns exactly this error, no need to catch it here

Reply:
Done :heavy_check_mark: